doc/ecmp/fine_grained_next_hop_hld.md

```diff
@@ -115,11 +120,14 @@ Warm restart is required for this design since the solution would benefit from h
 # 2 Modules Design
 
 ## 2.1 Config DB
-Following new tables will be added to Config DB. Unless otherwise stated, the attributes are mandatory.
+Following new tables will be added to Config DB. Unless otherwise stated, the attributes are mandatory. FG_NHG_MEMBER 
+table is not used or required for dynamic-nhg mode.
 ```
 FG_NHG|{{fg-nhg-group-name}}:
-    "bucket_size": {{hash_bucket_size}}
-    "match_mode" : {{route-based/nexthop-based}}
+    "bucket_size"   : {{hash_bucket_size}}
+    "match_mode"    : {{route-based/nexthop-based}}
+    "nhg_mode:      : {{static-nhg/dynamic-nhg}} (Optional)

```

Comment:
can we instead re-use match_mode itself?
route-based = filter by full route, prefix + nexthop IP
nexthop-based = filter by next-hop. All prefixes which match the next-hop IPs get fine grained ecmp.
prefix-based = filter by ip prefix in the route. All prefixes which match the specified prefix get fine grained ecmp regardless of the next-hop IPs.

Reply:
Hi Anish, I was initially thinking along the same lines but didn't want to redefine the functionality associated with route-based in the current HLD for backword compatibility reasons. In the following excerpt from the current HLD, route-based in defined as matching on the prefix only
`route-based looks to prefix to filter routes, and uses fine grained ecmp when the route prefix matches the FG_NHG_PREFIX prefix.
`

Reply:
The proposal to reuse match_mode will not redefine functionality and will also not cause any backward compatibility issues, because a brand new prefix-based match_mode being defined which does not impact the other matchmodes.

The other reason to keep using match_mode and not define a nhg_mode is: nexthop-based is also a type of dynamic-nhg, it dynamically learns BGP routes and matches based on the next-hop IP.

Reply:
Hi Anish, I understand your point but we were tripped by the existing definition of route-based in the yang model.

`            enum route-based {
               description
                 "Fine grained ecmp is used when the route prefix matches the FG_NHG_PREFIX prefix";
            }
`
In both the existing use case and the new one added by us, route prefix needs to match FG_NHG_PREFIX. So, we added another attribute to differentiate between the existing case where the list of nexthop IPs is statically configured vs our enhancement where nexthops IPs don't need to be specified though config.

Reply:
@anish-n As you originally suggested, I've modified the proposal to add prefix-based as an another enum value of match-mode to for dynamically learnt nexthop IPs. Please let me know if you have any other comments.